x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java

```diff
@@ -198,9 +199,11 @@ protected static Batch<LogicalPlan> operators() {
             new PropagateEmptyRelation(),
             new ConvertStringToByteRef(),
             new FoldNull(),
+            new ReplaceAggregatesWithNull(),
             new SplitInWithFoldableValue(),
             new PropagateEvalFoldables(),
             new ConstantFolding(),
+            new SubstituteSurrogates(),
```

Comment:
`SubstituteSurrogates` is also executed in `Substitutions` batch from the LogicalPlanOptimizer, but in reality it shouldn't be part of that batch because it **relies on `ConstantFolding`, `FoldNull` and `PropagateEvalFoldables`** to be able to do its job properly.

This leads me to the same idea of `SurrogateExpression` being more an aid for foldables, rather than a surrogate expression for every kind of expression. The true (non-foldables) surrogate expression (like the one in `TOP` aggregate function) should only be used **after all the constants have been resolved**, propagated and replaced.

Removing `SubstituteSurrogates` from the `Substitutions` batch is currently not possible because `TranslateMetricsAggregate` needs it. This `metrics` specific rule is also the reason why `ReplaceStatsNestedExpressionWithEval` is executed twice in the `Substitutions` batch. One idea is to **investigate** and see if it's possible to change the implementation of `TranslateMetricsAggregate` so that is not forcing multiple uses of the same rule in that batch; maybe `TranslateMetricsAggregate` should be dealt with **like `inlinestats` deals with the split of plans**.